### PR TITLE
add array element at function

### PR DESCRIPTION
**Pull request by walterddr**
support `array_element_at`

### CODE REVIEW

File: pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapperTest.java [hunk 1]
```diff
@@ -865,6 +866,91 @@ public void testConcatStringTransformFunction() {
...
+    }
+    testTransformFunction(transformFunction, expectedValues);
+  }
+
+  @Test
+  public void testArrayElementAtString() {
+    Random rand = new Random(42);
+    int index = 1 + rand.nextInt(_mvColumnLength);
+    ExpressionContext expression = RequestContextUtils.getExpression(
+        String.format("array_element_at_string(%s, %d)", STRING_MV_COLUMN, index));
```
Comment (walterddr) [inline pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapperTest.java:942]:
```suggestion
        String.format("array_element_at_string(%s, %d)", STRING_MV_COLUMN, index + 1));
```

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java [hunk 2]
```diff
@@ -151,4 +151,29 @@ public static double[] arrayConcatDouble(double[] values1, double[] values2) {
...
   }
+
+  @ScalarFunction
+  public static int arrayElementAtInt(int[] arr, int idx) {
+    return arr[idx];
+  }
+
+  @ScalarFunction
+  public static long arrayElementAtLong(long[] arr, int idx) {
+    return arr[idx];
```
Comment (walterddr) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java:162]:
```suggestion
    return arr[idx - 1];
```

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java [hunk 3]
```diff
@@ -151,4 +151,29 @@ public static double[] arrayConcatDouble(double[] values1, double[] values2) {
...
+  }
+
+  @ScalarFunction
+  public static double arrayElementAtDouble(double[] arr, int idx) {
+    return arr[idx];
+  }
+
+  @ScalarFunction
+  public static String arrayElementAtString(String[] arr, int idx) {
+    return arr[idx];
```
Comment (walterddr) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java:177]:
```suggestion
    return arr[idx - 1];
```

File: pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapperTest.java [hunk 4]
```diff
@@ -865,6 +866,91 @@ public void testConcatStringTransformFunction() {
...
+    }
+    testTransformFunction(transformFunction, expectedValues);
+  }
+
+  @Test
+  public void testArrayElementAtDouble() {
+    Random rand = new Random(42);
+    int index = 1 + rand.nextInt(_mvColumnLength);
+    ExpressionContext expression = RequestContextUtils.getExpression(
+        String.format("array_element_at_double(%s, %d)", DOUBLE_MV_COLUMN, index));
```
Comment (walterddr) [inline pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapperTest.java:925]:
```suggestion
    int index = rand.nextInt(_mvColumnLength);
    ExpressionContext expression = RequestContextUtils.getExpression(
        String.format("array_element_at_double(%s, %d)", DOUBLE_MV_COLUMN, index + 1));
```

File: pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapperTest.java [hunk 5]
```diff
@@ -865,6 +866,91 @@ public void testConcatStringTransformFunction() {
...
+    }
+    testTransformFunction(transformFunction, expectedValues);
+  }
+
+  @Test
+  public void testArrayElementAtFloat() {
+    Random rand = new Random(42);
+    int index = 1 + rand.nextInt(_mvColumnLength);
+    ExpressionContext expression = RequestContextUtils.getExpression(
+        String.format("array_element_at_float(%s, %d)", FLOAT_MV_COLUMN, index));
```
Comment (walterddr) [inline pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapperTest.java:908]:
```suggestion
    int index = rand.nextInt(_mvColumnLength);
    ExpressionContext expression = RequestContextUtils.getExpression(
        String.format("array_element_at_float(%s, %d)", FLOAT_MV_COLUMN, index + 1));
```

File: pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapperTest.java [hunk 6]
```diff
@@ -865,6 +866,91 @@ public void testConcatStringTransformFunction() {
...
+    }
+    testTransformFunction(transformFunction, expectedValues);
+  }
+
+  @Test
+  public void testArrayElementAtLong() {
+    Random rand = new Random(42);
+    int index = 1 + rand.nextInt(_mvColumnLength);
+    ExpressionContext expression = RequestContextUtils.getExpression(
+        String.format("array_element_at_long(%s, %d)", LONG_MV_COLUMN, index));
```
Comment (walterddr) [inline pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapperTest.java:891]:
```suggestion
    int index = rand.nextInt(_mvColumnLength);
    ExpressionContext expression = RequestContextUtils.getExpression(
        String.format("array_element_at_long(%s, %d)", LONG_MV_COLUMN, index + 1));
```

File: pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapperTest.java [hunk 7]
```diff
@@ -865,6 +866,91 @@ public void testConcatStringTransformFunction() {
     testTransformFunctionMV(transformFunction, expectedValues);
   }
 
+  @Test
+  public void testArrayElementAtInt() {
+    Random rand = new Random(42);
+    int index = 1 + rand.nextInt(_mvColumnLength);
+    ExpressionContext expression = RequestContextUtils.getExpression(
+        String.format("array_element_at_int(%s, %d)", INT_MV_COLUMN, index));
```
Comment (walterddr) [inline pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapperTest.java:874]:
```suggestion
    int index = rand.nextInt(_mvColumnLength);
    ExpressionContext expression = RequestContextUtils.getExpression(
        String.format("array_element_at_int(%s, %d)", INT_MV_COLUMN, index + 1));
```

**Comment by vvivekiyer** [comment]
@walterddr  Based on my understanding, pinot does not guarantee the ordering of elements in the record of an MV column. So can you please help me understand what will be the value of providing this functionality?

**Comment by somandal** [comment]
@walterddr does this PR indicate that the list of MVs in a row cannot be reordered going forward? There is nothing preventing such reordering today and we've also told users that they shouldn't rely on the order of the elements in a given MV row.

File: pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/BaseTransformFunctionTest.java [hunk 8]
```diff
@@ -109,6 +109,7 @@ public abstract class BaseTransformFunctionTest {
 
   protected Map<String, DataSource> _dataSourceMap;
   protected ProjectionBlock _projectionBlock;
+  protected int _mvColumnLength;
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/BaseTransformFunctionTest.java:112]: (minor) `_numMVValues`

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java [hunk 9]
```diff
@@ -151,4 +151,29 @@ public static double[] arrayConcatDouble(double[] values1, double[] values2) {
   public static String[] arrayConcatString(String[] values1, String[] values2) {
     return ArrayUtils.addAll(values1, values2);
   }
+
+  @ScalarFunction
+  public static int arrayElementAtInt(int[] arr, int idx) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java:156]: We might want to support negative index (from the last element)

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java (hunk 9 above)
Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java:157]: SQL standard element access starts with 1, which is a bit confusing with negative indexing in most language. let's leave this out for now. so although this is a pretty common usage

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java [hunk 10]
```diff
@@ -151,4 +152,29 @@ public static double[] arrayConcatDouble(double[] values1, double[] values2) {
...
+    return idx > 0 && idx <= arr.length ? arr[idx - 1] : (int) DataSchema.ColumnDataType.INT.getNullPlaceholder();
```
Comment (nizarhejazi) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java:158]: @walterddr Pinot Strings are zero-based while Presto's are one-based. Does it make sense to support one-based arrays in Pinot and zero-based strings?
cc: @Jackie-Jiang

Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java:158]: Seems Postgres is also one-based. We might want to change the string functions to one-based as well. Currently `substr` is zero-based. @nizarhejazi Do you see other zero-based string functions?

Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java:158]: Hmm, seems there are other zero-based string functions such as `strpos`, `split_part` etc.

Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java:158]: also there's no such function as `substr` in postgres. only `substring` and it is one-based.

Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java:158]: `substr` is supported in presto, but is one-based: https://prestodb.io/docs/current/functions/string.html

Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/ArrayFunctions.java:158]: Fix split not on whole separator #9650

**Pull request merged**